### PR TITLE
Tests browser/network/pastebin

### PR DESCRIPTION
**Pull request by arsana7**
Hi, I've written some tests for browser.network.pastebin.
If there is anything I can improve in it, please just say a word.

[Review on Reviewable](https://reviewable.io/reviews/the-compiler/qutebrowser/1320)

### CODE REVIEW

**Comment by The-Compiler** [comment]
Thank you! I have some little style nitpicks, but other than that it's looking quite good!

- Since `pastebin.py` now has 100% coverage, please add it to `PERFECT_FILES` in `scripts/dev/check_coverage.py` (as a `(test_file, tested_file)` tuple).
- I don't think it makes sense to separate `DATA`/`DATA_NOPARENT`/etc. from the lines where the `@pytest.mark.parametrize` is on. It'd make sense if it's used in multiple places, but I don't think it does here.

I'd prefer something like:

```python
@pytest.mark.parametrize('data', [
    {
        "name": "XYZ",
        "title": "hello world",
        "text": "xyz. 123 \n 172ANB",
        "reply": "abc"
    },
    {
        "name": "the name",
        "title": "the title",
        "text": "some Text",
        "reply": "some parent"
    }
])
def test_paste_with_parent(data):
    # [...]
```

- You could use a [pytest fixture](http://pytest.org/latest/fixture.html) for the `PastebinClient` to move setup out of the test functions:

```python
@pytest.fixture
def pbclient():
    client = pastebin.PastebinClient()
    http_stub = HTTPPostStub()
    client._client = http_stub
    return client
```

Then your functions can have a `pbclient` argument where they'll recieve the client.

(I also think `pbclient` or so is a better name - I first got confused between `client` and `client._client` when reviewing)

- Also check the output of the [flake8](https://travis-ci.org/The-Compiler/qutebrowser/jobs/110977323) / [pylint](https://travis-ci.org/The-Compiler/qutebrowser/jobs/110977326) linters on Travis for some little style issues (checks for [pep8](https://www.python.org/dev/peps/pep-0008/) and other things)

**Comment by arsana7** [comment]
I've done some refactoring, the code should be better now.

**Comment by The-Compiler** [comment]
Thanks again! It looks pretty good now, but I probably won't get to do a final review/merge before Thursday (or maybe even Sunday), so please stay tuned :wink:

File: tests/unit/browser/network/test_pastebin.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+"""Tests for qutebrowser.browser.network."""
+
+import pytest
+from PyQt5.QtCore import QUrl
+
+from qutebrowser.browser.network import pastebin
+from qutebrowser.misc import httpclient
+
+
+class HTTPPostStub(httpclient.HTTPClient):
```
Comment (The-Compiler) [inline tests/unit/browser/network/test_pastebin.py]: Is there a reason you inherit from `httpclient.HTTPClient` here? I think it's not needed, and it's error prone (accidental network requests when forgetting to override something).

**Comment by The-Compiler** [comment]
I think `HTTPPostStub` should actually be a `QObject` so you can test the real signals:

Other than inheriting from `QObject` you'll need to take an optional `parent` argument and pass it through, and add the signal:

```python
class HTTPPostStub(QObject):

    """A stub class for HTTPClient.

    [...]
    """

    success = pyqtSignal(str)
    error = pyqtSignal(str)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.url = None
        self.data = None

    [...]
```

Then you can adjust the tests to actually emit the signal:

```python
def test_on_client_success(http, pbclient, qtbot):
    with qtbot.assertNotEmitted(pbclient.error):
        with qtbot.waitSignal(pbclient.success):
            pbclient._client.success.emit(http)
```

This will ensure e.g. the signature in the `pyqtSlot` decorators is right too.

Please also add a test which checks the `error` connection, i.e. emits `pbclient._client.error` and makes sure `pbclient.error` is emitted.

After that, this should be ready to merge - thanks!

**Comment by The-Compiler** [comment]
Hey @arsana7 - any update on this? If you prefer me to make these changes myself, just tell me, no problem!

**Comment by arsana7** [comment]
Hi
Sorry, I don't have too much time now, and I don't know when I will be able to get into this again, so if you'd like to make the changes yourself - please, feel free to do so.

**Comment by The-Compiler** [comment]
No problem - thanks for the follow-up!

**Comment by The-Compiler** [comment]
I just used the opportunity to [adjust the code a bit](https://github.com/The-Compiler/qutebrowser/commit/86ab33c558c5d0dcae605b53c0eb9ea1f3dead23) to make testing easier, and then [adjusted the tests](https://github.com/The-Compiler/qutebrowser/commit/f43cbed72f535bc9b7e71dcd6a741f86c6faf90d) like mentioned in my comment.

Thanks again for the PR, it was a good foundation to build upon!

**Pull request merged**